### PR TITLE
James 1727 JamesMailetContext should rely on Mailqueue to send eMails

### PR DESCRIPTION
**Pull request by chibenwa**

### CODE REVIEW

File: server/mailet/mailetcontainer-camel/src/main/java/org/apache/james/mailetcontainer/impl/JamesMailetContext.java [hunk 1]
```diff
@@ -417,24 +417,6 @@ public void sendMail(MailAddress sender, Collection<MailAddress> recipients, Mim
...
-     * </p>
-     * <p>
-     * The local field localDeliveryMailet will be removed when we remove the
-     * storeMail method.
-     * </p>
-     *
-     * @deprecated since 2.2.0 look at the LocalDelivery code to find out how to
-     *             do the local delivery.
-     */
-    public void storeMail(MailAddress sender, MailAddress recipient, MimeMessage msg) {
```
Comment (mbaechler) [inline server/mailet/mailetcontainer-camel/src/main/java/org/apache/james/mailetcontainer/impl/JamesMailetContext.java]: are you sure you want to remove the method ? it's already deprecated

Comment (chibenwa) [inline server/mailet/mailetcontainer-camel/src/main/java/org/apache/james/mailetcontainer/impl/JamesMailetContext.java]: I deleted the related commit

File: server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/AbstractRedirect.java [hunk 2]
```diff
@@ -1025,6 +1025,7 @@ protected void updateHeaders() throws MessagingException {
             setIsReply(newMail, isReply(originalMail), originalMail);
 
             newMail.getMessage().saveChanges();
+            newMail.removeAllAttributes();
```
Comment (aduprat) [inline server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/AbstractRedirect.java]: No tests?

File: server/container/guice/memory-guice/src/main/java/org/apache/james/modules/server/MemoryMailQueueFactory.java [hunk 3]
```diff
@@ -76,8 +79,11 @@ public void enQueue(Mail mail, long delay, TimeUnit unit) throws MailQueueExcept
         }
 
         @Override
-        public void enQueue(Mail mail) throws MailQueueException {
-            mailItems.addFirst(new MemoryMailQueueItem(mail));
+        public void enQueue(Mail mail) throws MailQueueException {try {
```
Comment (aduprat) [inline server/container/guice/memory-guice/src/main/java/org/apache/james/modules/server/MemoryMailQueueFactory.java]: indent

File: server/container/core/src/test/java/org/apache/james/core/MimeMessageCopyOnWriteProxyTest.java [hunk 4]
```diff
@@ -178,8 +177,86 @@ public void testMessageCloning3() throws Exception {
...
+        MimeMessage m = new MimeMessageCopyOnWriteProxy(new MimeMessage(Session.getDefaultInstance(new Properties(null))));
+        m.setText("CIPS");
+        MailImpl mail = new MailImpl("test", new MailAddress("test@test.com"), r, m);
+        MailImpl mail2 = new MailImpl("test", new MailAddress("test@test.com"), r, m);
+        assertTrue(isSameMimeMessage(m, mail.getMessage()));
+        assertTrue(isSameMimeMessage(m, mail2.getMessage()));
+        // change the message that should be not referenced by mail that has
+        // been disposed, so it should not clone it!
+        System.gc();
+        Thread.sleep(100);
```
Comment (aduprat) [inline server/container/core/src/test/java/org/apache/james/core/MimeMessageCopyOnWriteProxyTest.java]: You may use `Awaitility`

File: server/container/core/src/test/java/org/apache/james/core/ReferenceCountTest.java [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+
+package org.apache.james.core;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import org.apache.james.lifecycle.api.Disposable;
+import org.junit.Before;
+import org.junit.Test;
+
+public class ReferenceCountTest {
```
Comment (aduprat) [inline server/container/core/src/test/java/org/apache/james/core/ReferenceCountTest.java]: You are not checking that `ExceptionId` is thrown.

**Comment by aduprat** [comment]
:+1:

File: mailet/base/src/test/java/org/apache/mailet/base/test/FakeMailContext.java [hunk 6]
```diff
@@ -44,11 +49,17 @@
         private final MailAddress sender;
         private final Collection<MailAddress> recipients;
         private final MimeMessage msg;
+        private final List<String> attributeNames;
 
-        public SentMail(MailAddress sender, Collection<MailAddress> recipients, MimeMessage msg) {
+        public SentMail(MailAddress sender, Collection<MailAddress> recipients, MimeMessage msg, Iterator<String> attributeNames) {
```
Comment (mbaechler) [inline mailet/base/src/test/java/org/apache/mailet/base/test/FakeMailContext.java]: iterator ? it's a very unusual type for an argument

File: mailet/base/src/test/java/org/apache/mailet/base/test/FakeMailContext.java [hunk 7]
```diff
@@ -44,11 +49,17 @@
         private final MailAddress sender;
         private final Collection<MailAddress> recipients;
         private final MimeMessage msg;
+        private final List<String> attributeNames;
```
Comment (mbaechler) [inline mailet/base/src/test/java/org/apache/mailet/base/test/FakeMailContext.java]: is it related to attributes below ?

File: mailet/base/src/test/java/org/apache/mailet/base/test/FakeMailContext.java [hunk 8]
```diff
@@ -188,10 +199,6 @@ public void setAttribute(String name, Serializable object) {
         attributes.put(name,object);
     }
 
-    public void storeMail(MailAddress sender, MailAddress recipient, MimeMessage msg) throws MessagingException {
```
Comment (mbaechler) [inline mailet/base/src/test/java/org/apache/mailet/base/test/FakeMailContext.java]: why ?

File: server/container/core/src/main/java/org/apache/james/core/ReferenceCount.java [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+import com.google.common.base.Preconditions;
+
+public class ReferenceCount<Id, ExceptionId extends Exception> {
+
+    public interface CloningOperation<Id, ExceptionId extends Exception> {
+        Id clone(Id clonee) throws ExceptionId;
+    }
+
+    private Id reference;
+    private CloningOperation<Id, ExceptionId> cloneOperation;
```
Comment (mbaechler) [inline server/container/core/src/main/java/org/apache/james/core/ReferenceCount.java]: final

File: server/container/core/src/main/java/org/apache/james/core/ReferenceCount.java [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+
+import com.google.common.base.Preconditions;
+
+public class ReferenceCount<Id, ExceptionId extends Exception> {
+
+    public interface CloningOperation<Id, ExceptionId extends Exception> {
+        Id clone(Id clonee) throws ExceptionId;
+    }
+
+    private Id reference;
```
Comment (mbaechler) [inline server/container/core/src/main/java/org/apache/james/core/ReferenceCount.java]: final

File: server/container/core/src/main/java/org/apache/james/core/ReferenceCount.java [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+    public synchronized void decrementReferenceCount() {
+        Preconditions.checkState(count > 0, "Can not decrement a reference that is not superior to 0");
+        count--;
+        if (count == 0) {
+            LifecycleUtil.dispose(reference);
+            reference = null;
+        }
+    }
+
+    public synchronized int getReferenceCount() {
```
Comment (mbaechler) [inline server/container/core/src/main/java/org/apache/james/core/ReferenceCount.java]: synchronized looks useless here

File: server/container/core/src/main/java/org/apache/james/core/ReferenceCount.java [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.core;
+
+import org.apache.james.lifecycle.api.LifecycleUtil;
+
+import com.google.common.base.Preconditions;
+
+public class ReferenceCount<Id, ExceptionId extends Exception> {
```
Comment (mbaechler) [inline server/container/core/src/main/java/org/apache/james/core/ReferenceCount.java]: Why Id ?

File: server/container/core/src/main/java/org/apache/james/core/ReferenceCount.java [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+            LifecycleUtil.dispose(reference);
+            reference = null;
+        }
+    }
+
+    public synchronized int getReferenceCount() {
+        return count;
+    }
+
+    public synchronized Id getWrapped() {
```
Comment (mbaechler) [inline server/container/core/src/main/java/org/apache/james/core/ReferenceCount.java]: Once you gave the reference, the whole ReferenceCount instance become useless

File: server/container/core/src/main/java/org/apache/james/core/ReferenceCount.java [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+        Preconditions.checkState(count > 0, "Attend to read content of the ReferenceCount while reference should not be tracked");
```
Comment (mbaechler) [inline server/container/core/src/main/java/org/apache/james/core/ReferenceCount.java]: if count <= 0, you throw an exception ?
I don't really understand the message.

File: server/container/core/src/main/java/org/apache/james/core/ReferenceCount.java [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+        return reference;
+    }
+
+    public synchronized ReferenceCount<Id, ExceptionId> getNewReferenceForWriting() throws ExceptionId {
```
Comment (mbaechler) [inline server/container/core/src/main/java/org/apache/james/core/ReferenceCount.java]: it's not a new reference, it's a cloned object

File: server/container/core/src/main/java/org/apache/james/core/ReferenceCount.java [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+        Preconditions.checkState(count > 0, "Attend to write content of the ReferenceCount while reference should not be tracked");
+        if (count > 1) {
+            ReferenceCount<Id, ExceptionId> result = new ReferenceCount<Id, ExceptionId>(cloneOperation.clone(reference), cloneOperation);
+            decrementReferenceCount();
```
Comment (mbaechler) [inline server/container/core/src/main/java/org/apache/james/core/ReferenceCount.java]: really ? what makes you believe the owner of this reference won't use it anymore ?

File: server/container/core/src/test/java/org/apache/james/core/MimeMessageCopyOnWriteProxyTest.java [hunk 17]
```diff
@@ -29,96 +31,103 @@
...
-        assertTrue(isSameMimeMessage(m2.getMessage(), mail.getMessage()));
-        mail.getMessage().setText("new body");
-        mail.getMessage().saveChanges();
-        // test it is different after a write operation!
-        mail.getMessage().setSubject("new Subject");
-        assertTrue(!isSameMimeMessage(m2.getMessage(), mail.getMessage()));
-        LifecycleUtil.dispose(mail);
-        LifecycleUtil.dispose(m2);
-        LifecycleUtil.dispose(messageFromSources);
+    public void mailShouldWrapMimeMessageCopyOnWriteProxy() throws Exception {
```
Comment (mbaechler) [inline server/container/core/src/test/java/org/apache/james/core/MimeMessageCopyOnWriteProxyTest.java]: and it's not what you assert

File: server/container/core/src/test/java/org/apache/james/core/MimeMessageCopyOnWriteProxyTest.java [hunk 18]
```diff
@@ -29,96 +31,103 @@
...
-        LifecycleUtil.dispose(m2);
-        assertEquals(mm3, getWrappedMessage(m2clone.getMessage()));
-        // change the message that should be not referenced by m2 that has
-        // been disposed, so it should not clone it!
-        m2clone.getMessage().setSubject("new Subject 2");
-        m2clone.getMessage().setText("new Body 3");
-        assertTrue(isSameMimeMessage(m2clone.getMessage(), mm));
-        LifecycleUtil.dispose(mail);
-        LifecycleUtil.dispose(messageFromSources);
+    public void mailShouldWrapMimeMessage() throws Exception {
```
Comment (mbaechler) [inline server/container/core/src/test/java/org/apache/james/core/MimeMessageCopyOnWriteProxyTest.java]: what part of the test asserts that ?

File: server/container/core/src/test/java/org/apache/james/core/MimeMessageCopyOnWriteProxyTest.java [hunk 19]
```diff
@@ -29,96 +31,103 @@
...
+        assertThat(mail.getMessage()).isNotSameAs(mail2.getMessage());
+    }
+
+    @Test
+    public void wrappedMimeMessageCopyOnWriteProxyShouldBeTheOriginalOne() throws Exception {
+        MimeMessageCopyOnWriteProxy messageFromSources = (MimeMessageCopyOnWriteProxy) getMessageFromSources(content + sep + body);
+        MailImpl mail = new MailImpl("test", new MailAddress("test@test.com"), recipientList, messageFromSources);
+        MailImpl mail2 = (MailImpl) mail.duplicate();
+
+        assertThat(isSameMimeMessage(mail2.getMessage(), mail.getMessage())).isTrue();
```
Comment (mbaechler) [inline server/container/core/src/test/java/org/apache/james/core/MimeMessageCopyOnWriteProxyTest.java]: why don't you use assertj for the assertion ?

File: server/container/core/src/test/java/org/apache/james/core/MimeMessageCopyOnWriteProxyTest.java [hunk 20]
```diff
@@ -127,95 +136,130 @@ public void testMessageCloning2() throws Exception {
...
-        // been disposed, so it should not clone it!
-        System.gc();
-        Thread.sleep(100);
-        mail.getMessage().setSubject("new Subject 2");
+    public void originalMimeMessageShouldBeChangedIfNotShared() throws Exception {
+        // Given
+        MailImpl mail = new MailImpl("test", new MailAddress("test@test.com"), recipientList, mimeMessage);
+
+        // When
         mail.getMessage().setText("new Body 3");
```
Comment (mbaechler) [inline server/container/core/src/test/java/org/apache/james/core/MimeMessageCopyOnWriteProxyTest.java]: is setText synchronized or what does happen when you share it at the same time you change to content ?

File: server/container/core/src/main/java/org/apache/james/core/ReferenceCount.java (hunk 12 above)
Comment (mbaechler) [inline server/container/core/src/main/java/org/apache/james/core/ReferenceCount.java]: So anyone can ask for any decrement at any time, that's it ?
How about generating a Closable reference that actually call the ReferenceCount.decrement on close ?

File: mailet/base/src/test/java/org/apache/mailet/base/test/FakeMailContext.java (hunk 8 above)
Comment (chibenwa) [inline mailet/base/src/test/java/org/apache/mailet/base/test/FakeMailContext.java]: Not used.... Moreover it does not belong to the MailContext interface. And it generated a warning in intelliJ...

File: server/container/core/src/main/java/org/apache/james/core/ReferenceCount.java (hunk 10 above)
Comment (chibenwa) [inline server/container/core/src/main/java/org/apache/james/core/ReferenceCount.java]: No. We have a reference = null used below as a gc optimisation technic, that forbids us to be final.

File: server/container/core/src/main/java/org/apache/james/core/ReferenceCount.java (hunk 15 above)
Comment (chibenwa) [inline server/container/core/src/main/java/org/apache/james/core/ReferenceCount.java]: It is a new refererence holding a cloned object suitable for writing.

File: server/container/core/src/main/java/org/apache/james/core/ReferenceCount.java (hunk 12 above)
Comment (chibenwa) [inline server/container/core/src/main/java/org/apache/james/core/ReferenceCount.java]: Id serves two purposes : 
 - It is a more general kind of stuff than just horlding a MimeMessage.
 - Make it easier for me to test this class behavior (without relying on a Mime message)

File: server/container/core/src/test/java/org/apache/james/core/MimeMessageCopyOnWriteProxyTest.java (hunk 20 above)
Comment (chibenwa) [inline server/container/core/src/test/java/org/apache/james/core/MimeMessageCopyOnWriteProxyTest.java]: I don't understand this comment.

File: server/container/core/src/test/java/org/apache/james/core/MimeMessageCopyOnWriteProxyTest.java (hunk 19 above)
Comment (chibenwa) [inline server/container/core/src/test/java/org/apache/james/core/MimeMessageCopyOnWriteProxyTest.java]: Ok

File: server/container/core/src/test/java/org/apache/james/core/MimeMessageCopyOnWriteProxyTest.java (hunk 17 above)
Comment (chibenwa) [inline server/container/core/src/test/java/org/apache/james/core/MimeMessageCopyOnWriteProxyTest.java]: Ok

File: server/container/core/src/test/java/org/apache/james/core/MimeMessageCopyOnWriteProxyTest.java (hunk 18 above)
Comment (chibenwa) [inline server/container/core/src/test/java/org/apache/james/core/MimeMessageCopyOnWriteProxyTest.java]: Using a MimeMessageWrapper for cloning returns a different message. So it is kind of implied. I will make it more readable.

File: mailet/base/src/test/java/org/apache/mailet/base/test/FakeMailContext.java (hunk 7 above)
Comment (chibenwa) [inline mailet/base/src/test/java/org/apache/mailet/base/test/FakeMailContext.java]: No it is not. This attributes are the one of a mail.
The one belows are more like configuration properties for a mailet.
Sorry, I did not choose the domain vocabulary....

File: mailet/base/src/test/java/org/apache/mailet/base/test/FakeMailContext.java (hunk 6 above)
Comment (chibenwa) [inline mailet/base/src/test/java/org/apache/mailet/base/test/FakeMailContext.java]: Ok

File: server/container/core/src/main/java/org/apache/james/core/ReferenceCount.java (hunk 12 above)
Comment (chibenwa) [inline server/container/core/src/main/java/org/apache/james/core/ReferenceCount.java]: Be it in a method named dispose or in the method decrement anyone can call it anywhen.
But be it? At least the name will be more explicit.

**Comment by mbaechler** [comment]
don't we need to close this PR ?

**Comment by chibenwa** [comment]
Not really. Sending mails still directly uses the processors directly, not behind the queue.

**Comment by mbaechler** [comment]
:+1:

File: server/container/guice/memory-guice/src/main/java/org/apache/james/modules/server/MemoryMailQueueFactory.java [hunk 21]
```diff
@@ -77,7 +80,11 @@ public void enQueue(Mail mail, long delay, TimeUnit unit) throws MailQueueExcept
 
         @Override
         public void enQueue(Mail mail) throws MailQueueException {
-            mailItems.addFirst(new MemoryMailQueueItem(mail));
+            try {
+                mailItems.addFirst(new MemoryMailQueueItem(new MailImpl(mail)));
```
Comment (rouazana) [inline server/container/guice/memory-guice/src/main/java/org/apache/james/modules/server/MemoryMailQueueFactory.java]: You don't need to dispose your new MailImpl?

Comment (chibenwa) [inline server/container/guice/memory-guice/src/main/java/org/apache/james/modules/server/MemoryMailQueueFactory.java]: No.
Initial mail will be disposed in his scope. You have dispose instructions after enqueues (hence my troubles)
The copied mail will be disposed after his upcoming processing

Comment (chibenwa) [inline server/container/guice/memory-guice/src/main/java/org/apache/james/modules/server/MemoryMailQueueFactory.java]: Check ActiveMQMailQueue and JMSMailQueue, they do not dispose the enqueued mail too.

The calling scope is responsible for this.

Comment (rouazana) [inline server/container/guice/memory-guice/src/main/java/org/apache/james/modules/server/MemoryMailQueueFactory.java]: Sorry, not sure to understand you answer. As you're duplicating the resource, are you sure there are 2 callings scopes that will dispose it?

Comment (chibenwa) [inline server/container/guice/memory-guice/src/main/java/org/apache/james/modules/server/MemoryMailQueueFactory.java]: The mail passed by the argument of the method will be disposed by the caller
The mail enqueued will be disposed by the caller of dequeue

Is that more clear ?

Comment (rouazana) [inline server/container/guice/memory-guice/src/main/java/org/apache/james/modules/server/MemoryMailQueueFactory.java]: Yes, thanks.

**Comment by rouazana** [comment]
:+1:

**Pull request merged**